llvm/lib/Transforms/IPO/SampleProfile.cpp

```diff
@@ -2443,53 +2372,222 @@ void SampleProfileMatcher::runOnFunction(const Function &F) {
+  auto &MismatchedCallsites =
+      FuncCallsiteMatchStates[FunctionSamples::getCanonicalFnName(F.getName())];
+
+  auto MapIRLocToProfileLoc = [&](const LineLocation &IRLoc) {
+    const auto &ProfileLoc = IRToProfileLocationMap.find(IRLoc);
+    if (ProfileLoc != IRToProfileLocationMap.end())
+      return ProfileLoc->second;
+    else
+      return IRLoc;
+  };
+
+  std::set<LineLocation> MatchedCallsites;
+  for (const auto &I : IRAnchors) {
+    // In post-match, use the matching result to remap the current IR callsite.
+    const auto &Loc = MapIRLocToProfileLoc(I.first);
+    const auto &IRCalleeName = I.second;
+    const auto &It = ProfileAnchors.find(Loc);
+    if (It == ProfileAnchors.end())
       continue;
-    runOnFunction(F);
+    const auto &Callees = It->second;
+
+    // Since indirect call does not have CalleeName, check conservatively if
+    // callsite in the profile is a callsite location. This is to reduce num of
+    // false positive since otherwise all the indirect call samples will be
+    // reported as mismatching.
+    if (IRCalleeName == SampleProfileMatcher::UnknownIndirectCallee)
+      MatchedCallsites.insert(Loc);
+    // TODO : Ideally, we should ensure it's a direct callsite location(Callees
+    // size is 1). However, there may be a bug for profile merge(like ODR
+    // violation) that causes the callees size to be more than 1. After we fix
+    // the bug, we can remove this check.
+    else if (Callees.count(getRepInFormat(IRCalleeName)))
+      MatchedCallsites.insert(Loc);
+  }
+
+  // Check if there are any callsites in the profile that does not match to any
+  // IR callsites, those callsite samples will be discarded.
+  for (const auto &I : ProfileAnchors) {
+    const auto &Loc = I.first;
+    [[maybe_unused]] const auto &Callees = I.second;
+    assert(!Callees.empty() && "Callees should not be empty");
+    if (IsPostMatch) {
+      if (MatchedCallsites.count(Loc)) {
+        auto It = MismatchedCallsites.find(Loc);
+        if (It != MismatchedCallsites.end() &&
+            It->second == MatchState::Mismatched)
+          MismatchedCallsites.emplace(Loc, MatchState::Recovered);
+      } else
+        MismatchedCallsites.emplace(Loc, MatchState::Mismatched);
+    } else {
+      if (MatchedCallsites.count(Loc))
+        MismatchedCallsites.emplace(Loc, MatchState::Matched);
+      else
+        MismatchedCallsites.emplace(Loc, MatchState::Mismatched);
+    }
+  }
+}
+
+void SampleProfileMatcher::countMismatchedFuncSamples(
+    const FunctionSamples &FS) {
+  const auto *FuncDesc = ProbeManager->getDesc(FS.getGUID());
+  // Skip the function that is external or renamed.
+  if (!FuncDesc)
+    return;
+
+  if (ProbeManager->profileIsHashMismatched(*FuncDesc, FS)) {
+    MismatchedFunctionSamples += FS.getTotalSamples();
+    return;
+  }
+  for (const auto &I : FS.getCallsiteSamples())
+    for (const auto &CS : I.second)
+      countMismatchedFuncSamples(CS.second);
```

Comment:
This is worth a comment (not about what it's doing but why). 

if inlinee's standalone function hash mismatches, do we not load inlinee's proifle? I thought we don't check hash for inlinees and always load their profile. Though if there is mismatch the load may not be accurate.

Reply:
> This is worth a comment (not about what it's doing but why).
> 
> if inlinee's standalone function hash mismatches, do we not load inlinee's proifle? I thought we don't check hash for inlinees and always load their profile. Though if there is mismatch the load may not be accurate.

Yeah, it should be a bug/inconsistency before, before `SalvageStaleProfile` is enabled, we only skip loading the top-level checksum mismatched function but not the inlinee's. Now with `SalvageStaleProfile` on, this is consistent, should not be a problem

Reply:
> Now with SalvageStaleProfile on, this is consistent, should not be a problem 

are you saying that now we also skip loading inlinee if inlinee's outline function's checksum mismatches? I don't see it doing that (FunctionSamples::findFunctionSamplesAt doesn't bother checksum). Can you share code pointer?

Reply:
> > Now with SalvageStaleProfile on, this is consistent, should not be a problem
> 
> are you saying that now we also skip loading inlinee if inlinee's outline function's checksum mismatches? I don't see it doing that (FunctionSamples::findFunctionSamplesAt doesn't bother checksum). Can you share code pointer?

Oh, I meant the opposite way, now with `SalvageStaleProfile` on, we load both the outline and inline function. (as https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/IPO/SampleProfile.cpp#L1838 return false)

here I wanted to say: we do load a checksum-mismatched function profile. but as it's checksum-mismatched, all the children inlinees are likely mismatched and dropped, so we conservatively count the total_samples as `MismatchedFunctionSamples`

Reply:
Strictly speaking,  counting the whole samples(`total_samples`) as mismatches makes sense in the current condition that all the call probe is after BB blocks, as one bb block changes all the call probe are mismatched. After we fix this, it would become less representative.. this is more like how many samples are processed by the sample profile matching.